raft/raft.go

```diff
@@ -0,0 +1,403 @@
+	config := raft.DefaultConfig()
+
+	var err error
+	nodeID := raftConfig.NodeID
+	if raftConfig.NodeID == "" {
+		nodeID, err = os.Hostname()
+		if err != nil {
+			return nil, fmt.Errorf("error getting hostname: %w", err)
+		}
+	}
+	raftAddr := raftConfig.Address
+	config.LocalID = raft.ServerID(nodeID)
+	raftDir := filepath.Join(raftConfig.Directory, nodeID)
+	err = os.MkdirAll(raftDir, os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("error creating raft directory: %w", err)
+	}
+
+	// Create the FSM
+	fsm := NewFSM()
+
+	// Create the log store and stable store
+	logStore, err := raftboltdb.NewBoltStore(filepath.Join(raftDir, "raft-log.db"))
+	if err != nil {
+		return nil, fmt.Errorf("error creating log store: %w", err)
+	}
+
+	stableStore, err := raftboltdb.NewBoltStore(filepath.Join(raftDir, "raft-stable.db"))
+	if err != nil {
+		return nil, fmt.Errorf("error creating stable store: %w", err)
+	}
+
+	// Create the snapshot store
+	snapshotStore, err := raft.NewFileSnapshotStore(raftDir, maxSnapshots, os.Stderr)
+	if err != nil {
+		return nil, fmt.Errorf("error creating snapshot store: %w", err)
+	}
+
+	// Setup Raft communication
+	addr, err := net.ResolveTCPAddr("tcp", raftAddr)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving TCP address: %w", err)
+	}
+	transport, err := raft.NewTCPTransport(raftAddr, addr, maxPool, transportTimeout, os.Stderr)
+	if err != nil {
+		return nil, fmt.Errorf("error creating TCP transport: %w", err)
+	}
+
+	// Create the Raft node
+	raftNode, err := raft.NewRaft(config, fsm, logStore, stableStore, snapshotStore, transport)
+	if err != nil {
+		return nil, fmt.Errorf("error creating Raft: %w", err)
+	}
+
+	node := &Node{
+		raft:          raftNode,
+		config:        config,
+		Fsm:           fsm,
+		logStore:      logStore,
+		stableStore:   stableStore,
+		snapshotStore: snapshotStore,
+		transport:     transport,
+		Logger:        logger,
+		Peers:         raftConfig.Peers,
+		grpcAddr:      raftConfig.GRPCAddress,
+	}
+
+	// Initialize RPC client
+	node.rpcClient = newRPCClient(node)
+
+	// Start RPC server
+	if err := node.startRPCServer(); err != nil {
+		return nil, fmt.Errorf("failed to start RPC server: %w", err)
+	}
+
+	// Handle bootstrapping
+	if raftConfig.IsBootstrap {
+		configuration := raft.Configuration{
+			Servers: make([]raft.Server, len(node.Peers)),
+		}
+		for i, peer := range node.Peers {
+			configuration.Servers[i] = raft.Server{
+				ID:      raft.ServerID(peer.ID),
+				Address: raft.ServerAddress(peer.Address),
+			}
+		}
+		configuration.Servers = append(configuration.Servers, raft.Server{
+			ID:      config.LocalID,
+			Address: transport.LocalAddr(),
+		})
+		node.raft.BootstrapCluster(configuration)
+	}
+
+	go node.monitorLeadership()
+
+	return node, nil
+}
+
+// monitorLeadership checks if the node is the Raft leader and logs state changes.
+func (n *Node) monitorLeadership() {
+	for {
+		isLeader := n.raft.State() == raft.Leader
+		if isLeader {
```

Comment:
```suggestion
		if n.raft.State() == raft.Leader {
```

Reply:
Addressed in [586efd7](https://github.com/gatewayd-io/gatewayd/pull/636/commits/586efd79aa8e19865cf1600ae2c948c64350188c)